<a href="https://colab.research.google.com/github/cchan555/lis4693/blob/main/lab-2/ChanLab2Feb242026.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [4]:
from requests.models import Response
!pip install spacy
!python -m spacy download en_core_web_sm
import spacy
import requests

# Load in and read file from GitHub
url = "https://raw.githubusercontent.com/cchan555/lis4693/refs/heads/main/lab-1/transcript"
response = requests.get(url)
response.raise_for_status() # Raise exception for HTTP errors
text = response.text

print("Number of Characters:", len(text)) # Print the number of characters
print(text[:300]) # Print the first 300 characters

# Sentence Segmentation
nlp = spacy.blank("en")
nlp.add_pipe("sentencizer")
doc = nlp(text)
sentences = list(doc.sents)
print("Number of sentences:", len(sentences))
print("\nFirst 3 sentences:")
for sent in sentences[:3]:
  print(sent)

# Word count and Token Analysis
words = [token.text.lower() for token in doc if token.is_alpha]
print("Total Words:", len(words))
print("Unique Words:", len(set(words)))

# Most Frequent Words
from collections import Counter
word_freq = Counter(words)
print("Top 10 most frequent words:")
for word, count in word_freq.most_common(10):
  print(word, count)

# Use Full spaCy Pipeline
nlp2 = spacy.load("en_core_web_sm")
doc2 = nlp2(text)
print("Named Entities:")
for ent in doc2.ents:
  print(ent.text, ent.label_)
print("Total number of Entities:", len(doc2.ents))

# PhraseMatcher
from spacy.matcher import PhraseMatcher
matcher = PhraseMatcher(nlp2.vocab, attr = "LOWER")
phrases = ["snowball", "county", "oklahoma"]
patterns = [nlp2(p) for p in phrases]
matcher.add("Terms", patterns)
matches = matcher(doc2)
print("Matches found:")
for match_id, start, end in matches:
  print(doc2[start:end])
  print("Sentence:", doc2[start].sent)


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.8/12.8 MB 104.9 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.
Number of Characters: 7240
REPORTING IN WASHINGTON.
I’M JACKIE DEFUSCO.
WELL, WE’VE BEEN WATCHING THE
RADAR A LOT OVER THE LAST COUPLE
OF DAYS.
AND SO HOW MUCH SNOW HAS FALLEN?
WELL, IF YOU LOOK HERE, YOU CAN
SEE THERE’S BEEN A COUPLE BANDS
OUT ACROSS NORTHERN OKLAHOMA.
AND THEN JUST SOUTH OF I-40, WE
HAD SOME SNOW LAST NIGHT
Number of sentences: 117

First 3 sentences:
REPORTING IN WASHINGTON.

I’M JACKIE DEFUSCO.

WELL, WE’VE BEEN WATCHING THE
RADAR A LOT OVER THE LAST COUPLE
OF DAYS.
Total Words: 1323
Unique Words: 391
Top 10 most frequent words:
to 47
and 46
the 44
is 29
s

Task 1: Completed <br>
Task 2: Completed <br>
Task 3: Perform sentences segmentation using the blank pipeline <br>
*   How many sentenses are in your transcript? <br> 117
*   Print the first 3 sentences: (The 3 sentences are also outputted in the console above) <br>
REPORTING IN WASHINGTON. <br>
I’M JACKIE DEFUSCO. <br>
WELL, WE’VE BEEN WATCHING THE
RADAR A LOT OVER THE LAST COUPLE
OF DAYS. <br>

Task 4: Perform Word Count & Token Analysis
*   What is the total number of words? <br> 1323
*   What is the number of unique words? <br> 391

Task 5: Find Most Frequent Words
*   What is the most frequent word? <br>
The most frequent word in this transcript is "to" with 47 times appearing.
*   Why do you think this word appears frequently? <br>
These words appeared frequenly because it is a common transition word to use. The work "to" is commonly used in English and explaining a topic. In this case, the video was about a news reporter explaining the weather situation in Oklahoma. For example, the new reporter talks about snow moving one place to another. The transition between the two places will most likely use the word "to".

Task 6: Run full spaCy pipeline
*   How many named entities were found? <br>
There at 95 entities found in this transcript.
*   What type of entities appear? (PERSON, ORG, DATE, etc.) <br>
The types of entities that appeared were the following: ORG, GPE, LOC, PERSON, CARDINAL, QUANTITY, TIME, MONEY, PRODUCT, WORK_OF_ART, NORP, and DATE.

Task 7: Use PhaseMatcher
*   What topic or subject were you interested in? <br>
This topic interested me because it was related to snow. The snow fall broke the all time high and I was doing research into snow fall at the time. At this time, Oklahoma schools had closing or virtual days, so I had more time to research into the history of snowfall and watch the news about it.

*   Identitfy three speciifc phrases that are directly relevant to search: <br>
The three phrases I chose that are related to my search are snowball, county, and oklahoma.

Task 8: Brief Reflection:
*   What went well? <br>
During lab 2, what went well was the step by step instructions that guided me through the process. The detailed documentation of the the new library spaCy was usefull as well becauase it was my first time being exposeed to this library. I also liked that going towards the end, I would apply some of the things that were shown in the previous step to the next steps. For example, when identifying the named entities, I use the print() format that was shown to me in a previous step to this step so I did not have to count how many entities there were.

*   What did not go well or what challenges you encountered? <br>
There were not many majory challanges that I encountered during this lab. Each step was explained with detail and there were a lot of resouces to help me if I needed help. The only issue that I remember encountering was during the load and read phase of GitHub. I did not know how to upload the link properly until messing around with GitHub and finding the "Raw" button.


